smarts/core/smarts.py

```diff
@@ -735,7 +736,6 @@ def _nondynamic_provider_step(
 
             vehicle = self._vehicle_index.vehicle_by_id(vehicle_id)
             assert isinstance(vehicle.chassis, BoxChassis)
-            vehicle.step(self._elapsed_sim_time)
```

Comment:
I am not entirely convinced it is reasonable to move `vehicle.step()` to after generating the `VehicleState`. It would appear to me that the most recent state should rely on a step having been performed.

Reply:
The reason I moved it is because it needs to happen whenever `pybullet` is stepped and having it happen in different locations runs the risk of it being missed.  Calling it is now a stronger "requirement" because the "step cache" in `chassis.py` needs to be cleared after each step, and this is the way to trigger that.

One effect of moving it into the `_step_pybullet()` function is to make it happen slightly earlier than it would have before.  Before, we waited to call it until after collisions were processed, but this does not seem to be problematic.  (Note that I think that, even before, it has always been called before generating the `VehicleState` object.)

However, another effect of moving it is that it is now called for _every_ vehicle, as opposed to just those with agents controlling them (due to moving it to before the `continue` in `_pybullet_provider_step()`).  I looked into this when I did it and it didn't seem problematic  to me, but I definitely think this decision should be reviewed.  (The `_has_stepped` state in `vehicle.py` just seems to be for controlling a warning message.)